Permissions audit

ThatGalNatalie opened the pull request:
PR Checklist:

- [x] add a short description of what's changed to the top of the `CHANGELOG.md`
- [ ] add/update tests (or don't, for reasons explained below)

## Describe this PR
Replaces everywhere we are using `this.tenant.role` with the `hasPermission` check

znicholasbrown commented on src/pages/TeamSettings/Account/Billing.vue:
```diff
@@ -38,7 +38,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this check should be on `update:license`

znicholasbrown commented on src/pages/TeamSettings/Account/License.vue:
```diff
@@ -21,10 +21,15 @@ export default {
     ...mapGetters('tenant', ['tenant']),
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I don't think this is used in this component but probably should be changed to `update:license` and used to disable the cancel plan button

znicholasbrown commented on src/pages/TeamSettings/CloudHooks.vue:
```diff
@@ -78,8 +78,12 @@ export default {
     isReadOnlyUser() {
-      return this.role === 'READ_ONLY_USER'
+      return (
+        !this.hasPermission('create', 'role') &&
+        !this.hasPermission('create', 'cloud-hook')
```
What's the purpose of the `create:role` check here? We'll want to break out the cloud hook CRUD permissions based on the operations the user can take, i.e. `create:cloud-hook` for creating new ones, create and delete for editing (since this isn't a true edit/update) etc

znicholasbrown commented on src/pages/TeamSettings/FlowConcurrency.vue:
```diff
@@ -119,7 +119,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this check should be `create:concurrency-limit`, `update:concurrency-limit`, and `delete:concurrency-limit`

znicholasbrown commented on src/pages/TeamSettings/FlowGroups.vue:
```diff
@@ -112,7 +113,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this check needs to be the same as the flow actions checks

znicholasbrown commented on src/pages/TeamSettings/Members.vue:
```diff
@@ -90,7 +90,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
This should be broken up to `create:membership`, `update:membership`, and `delete:membership` checks

znicholasbrown commented on src/pages/TeamSettings/Projects.vue:
```diff
@@ -120,10 +121,10 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
What's the purpose of the `create:role` check here?

znicholasbrown commented on src/pages/TeamSettings/Projects.vue:
```diff
@@ -120,10 +121,10 @@ export default {
     },
     isReadOnlyUser() {
-      return this.tenant.role === 'READ_ONLY_USER'
+      return !this.isTenantAdmin && !this.hasPermission('create', 'project')
```
This should be broken up to `create:project`, `update:project`, and `delete:project` checks

znicholasbrown commented on src/pages/TeamSettings/Secrets.vue:
```diff
@@ -90,8 +90,13 @@ export default {
     isReadOnlyUser() {
-      return this.role === 'READ_ONLY_USER'
+      return (
+        !this.hasPermission('create', 'role') &&
```
What's the purpose of the `create:role` check here?

znicholasbrown commented on src/pages/TeamSettings/Service-Accounts.vue:
```diff
@@ -41,9 +41,9 @@ export default {
+    ...mapGetters('license', ['license', 'hasPermission']),
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this should be doing individual checks on create, update, and delete `*:service-account`

znicholasbrown commented on src/pages/TeamSettings/TaskConcurrency.vue:
```diff
@@ -130,7 +130,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this check should be the same as the flow concurrency

znicholasbrown commented on src/pages/TeamSettings/TeamSettings.vue:
```diff
@@ -27,7 +27,11 @@ export default {
-    ...mapGetters('tenant', ['tenant', 'role'])
+    ...mapGetters('tenant', ['tenant', 'role']),
+    ...mapGetters('license', ['hasPermission']),
+    isTenantAdmin() {
```
This probably isn't necessary since this button will be removed soon and isn't visible anyway

znicholasbrown commented on src/pages/TeamSettings/Tokens.vue:
```diff
@@ -67,11 +67,12 @@ export default {
     },
     isTenantAdmin() {
-      return this.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
What's the purpose of the `create:role` check here?

znicholasbrown commented on src/components/ScheduleToggle.vue:
```diff
@@ -20,11 +20,15 @@ export default {
     isReadOnlyUser() {
-      return this.role === 'READ_ONLY_USER'
+      return (
+        !this.hasPermissions('create', 'role') &&
+        !this.hasPermissions('create', 'flow')
```
I think we need to check 2 things here: some users will have `create:run`  permission but not `delete:run` (and vice versa) so the quick run button will need to be gated on `create:run` and the schedule toggle will need to be gated for both scenarios; when a user has `create:run` but not `delete:run` they should be able to turn a schedule on but not off etc (same comment as below)

znicholasbrown reviewed (dismissed):
Lots of comments on this one, it looks like you've added `create:role` checks in a lot of places as a stand in for a tenant admin role check, which I don't think is correct (I could be missing some context here). More broadly we should be removing references to things like read only users and tenant admins in favor of using the more granular permissions

ThatGalNatalie commented on src/components/CancelButton.vue:
```diff
@@ -1,8 +1,18 @@
+    ...mapGetters('license', ['hasPermission']),
+    isReadOnly() {
+      return (
+        !this.hasPermission('create', 'role') &&
+        !this.hasPermission('create', 'flow')
```
For the `create:role` check, I was trying to differentiate between all the roles. Here, I guess I was checking if they're not a tenant admin

ThatGalNatalie commented on src/components/DeleteFlowButton.vue:
```diff
@@ -34,6 +34,13 @@ export default {
+    ...mapGetters('license', ['hasPermission']),
+    isReadOnly() {
+      return (
+        !this.hasPermission('create', 'role') &&
+        !this.hasPermission('delete', 'flow')
```
The `create:role` permission check here was to differentiate between all the roles and determine that this is read only. Now, it makes more sense to not include that check.

ThatGalNatalie commented on src/components/LabelEdit.vue:
```diff
@@ -47,6 +47,13 @@ export default {
+    ...mapGetters('license', ['hasPermission']),
+    isReadOnly() {
+      return (
+        !this.hasPermission('create', 'role') &&
+        !this.hasPermission('create', 'flow')
```
The `create:role` permission check here was to differentiate between all the roles and determine that this is read only. I think the `update:run` makes more sense.

ThatGalNatalie commented on src/pages/Dashboard/NewProject-Dialog.vue:
```diff
@@ -28,7 +28,15 @@ export default {
+
+    isReadOnly() {
+      return (
+        !this.hasPermission('create', 'role') &&
```
The `create:role` permission check here was to differentiate between all the roles and determine that this is read only. Makes more sense to just check `create:project`

zhen0 commented on src/components/License/ChangePlanDialog.vue:
```diff
@@ -21,13 +21,13 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
I think this will be license related permission not role?

zhen0 commented on src/components/License/UpgradeAlert.vue:
```diff
@@ -7,16 +7,15 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'license')
```
Not sure create is correct here.  Possibly update and/or read.

ThatGalNatalie replied:
This is currently not being used in this component. Should we delete it?

zhen0 commented on src/components/ScheduleToggle.vue:
```diff
@@ -118,7 +116,11 @@ export default {
               :loading="loading"
-              :disabled="isReadOnlyUser || archived"
+              :disabled="
+                !hasPermission('create', 'run') ||
```
If they don't have create run permission but do have delete run they should still be able to turn the schedule off

zhen0 commented on src/pages/FlowRun/Actions.vue:
```diff
@@ -233,7 +234,7 @@ export default {
       </template>
       <span v-if="isReadOnlyUser">
-        Read-only users cannot delete flows
+        You don't have permission to delete flows
```
This will need delete:flow permissions?

ThatGalNatalie replied:
I was thinking delete:run since it's trying to delete a flow run?

zhen0 commented on src/pages/Plans.vue:
```diff
@@ -75,7 +74,7 @@ export default {
     },
     isTenantAdmin() {
-      return this.tenant.role === 'TENANT_ADMIN'
+      return this.hasPermission('create', 'role')
```
Check this - Plans page is not about roles.

zhen0 reviewed (commented):
A few more comments/checks from me.  Realize this may still be a work in progress.  We'll need to audit this one thoroughly to make sure we get all those permissions correct.  I've made some suggestions for what permissions should be but please reach out in Cloud-UI channel / Cloud channel to make sure they're correct.

zhen0 commented:
Thanks for the changes @ThatGalNatalie. Please can we merge this dev so we can QA without affecting staging.

Also one tiny extra nitpick from me - can we switch 'isReadOnly' and 'isTenantAdmin' role names to something like 'incorrectPermissions' or similar to keep our code up to date with changes.

ThatGalNatalie commented:
> Thanks for the changes @ThatGalNatalie. Please can we merge this dev so we can QA without affecting staging.
> 
> Also one tiny extra nitpick from me - can we switch 'isReadOnly' and 'isTenantAdmin' role names to something like 'incorrectPermissions' or similar to keep our code up to date with changes.

Updated all references of isReadOnly and isTenantAdmin to permissionsCheck and switched the base branch to dev

zhen0 commented:
Noting that we may want to re-open against master at some point in the future so restoring the branch.
